src/Groups/matrices/matrix_manipulation.jl

```diff
@@ -0,0 +1,260 @@
+      C[i+s-1,j+t-1] = B[s,t]
+   end
+   end
+   return C
+end
+
+"""
+    insert_block!(A::MatElem, B::MatElem, i,j)
+Insert the block `B` in the matrix `A` at the position `(i,j)`.
+"""
+function insert_block!(A::MatElem{T}, B::MatElem{T}, i::Int, j::Int) where T <: RingElem
+   for s in 1:nrows(B)
+   for t in 1:ncols(B)
+      A[i+s-1,j+t-1] = B[s,t]
+   end
+   end
+   return A
+end
+
+# exists already in Hecke cat(V...; dims=(1,2))
+"""
+    diagonal_join(V::AbstractVector{<:MatElem})
+    diagonal_join(V::T...) where T <: MatElem
+Return the diagonal join of the matrices in `V`.
+"""
+function diagonal_join(V::AbstractVector{T}) where T <: MatElem
+   nr = sum([nrows(v) for v in V])
+   nc = sum([ncols(v) for v in V])
+   B = zero_matrix(base_ring(V[1]), nr,nc)
+   pos_i=1
+   pos_j=1
+   for k in 1:length(V)
+      insert_block!(B,V[k],pos_i,pos_j)
+      pos_i += nrows(V[k])
+      pos_j += ncols(V[k])
+   end
+   return B
+end
+
+diagonal_join(V::T...) where T <: MatElem = diagonal_join(collect(V))
+
+#=
+diagonal_join(V::T...) where T <: MatElem = cat(V; dims=(1,2))
+=#
+
+"""
+    block_matrix(m::Int, n::Int, V::AbstractVector{T}) where T <: MatElem
+Return the matrix constructed from the given block matrices, which should be given as a sequence `V` of `m*n` matrices (given in row major order, in other words listed across rows). 
+"""
+function block_matrix(m::Int, n::Int, V::AbstractVector{T}) where T <: MatElem
+   length(V)==m*n || throw(ArgumentError("Wrong number of inserted blocks"))
+   n_rows=0
+   for i in 1:m
+      for j in 1:n
+         nrows(V[n*(i-1)+j])==nrows(V[n*(i-1)+1]) || throw(ArgumentError("Invalid matrix dimension"))
+         ncols(V[n*(i-1)+j])==ncols(V[j]) || throw(ArgumentError("Invalid matrix dimension"))
+      end
+      n_rows += nrows(V[n*(i-1)+1])
+   end
+   n_cols = sum([ncols(V[j]) for j in 1:n])
+   B = zero_matrix(base_ring(V[1]), n_rows, n_cols)
+   pos_i=1
+   for i in 1:m
+      pos_j=1
+      for j in 1:n
+         insert_block!(B,V[n*(i-1)+j],pos_i,pos_j)
+         pos_j += ncols(V[n*(i-1)+j])
+      end
+      pos_i += nrows(V[n*(i-1)+1])
+   end
+   return B
+end
+
+"""
+    matrix(A::Array{AbstractAlgebra.Generic.FreeModuleElem{T},1})
+Return the matrix whose rows are the vectors in `A`. Of course, vectors in `A` must have the same length and the same base ring.
+"""
+function matrix(A::Array{AbstractAlgebra.Generic.FreeModuleElem{T},1}) where T <: FieldElem
+   c = length(A[1].v)
+   X = zero_matrix(base_ring(A[1]), length(A), c)
+   for i in 1:length(A)
+      @assert length(A[i].v)==c "Vectors must have the same length"
+      for j in 1:c X[i,j] = A[i][j] end
+   end
+
+   return X
+end
+
+"""
+    conjugate_transpose(x::MatElem{T}) where T <: FieldElem
+If the base ring of `x` is `GF(q^2)`, return the matrix `transpose( map ( y -> y^q, x) )`. An error is returned if the base ring has no even degree.
+"""
+function conjugate_transpose(x::MatElem{T}) where T <: FieldElem
```

Comment:
type should indicate finite field (please look for other places that may need this changed)